subsys/bluetooth/mesh/access.c

```diff
@@ -812,80 +810,71 @@ const struct bt_mesh_comp *bt_mesh_comp_get(void)
 	return dev_comp;
 }
 
-struct bt_mesh_model *bt_mesh_model_root(struct bt_mesh_model *mod)
+void bt_mesh_model_extensions_walk(struct bt_mesh_model *model,
+				   enum bt_mesh_walk (*cb)(struct bt_mesh_model *mod,
+							   void *user_data),
+				   void *user_data)
 {
-#ifdef CONFIG_BT_MESH_MODEL_EXTENSIONS
-	while (mod->next) {
-		mod = mod->next;
+#ifndef CONFIG_BT_MESH_MODEL_EXTENSIONS
+	(void)cb(model, user_data);
+	return;
+#else
+	struct bt_mesh_model *it;
+
+	if (model->next == NULL) {
+		(void)cb(model, user_data);
+		return;
 	}
-#endif
-	return mod;
-}
-
-void bt_mesh_model_tree_walk(struct bt_mesh_model *root,
-			     enum bt_mesh_walk (*cb)(struct bt_mesh_model *mod,
-						     uint32_t depth,
-						     void *user_data),
-			     void *user_data)
-{
-	struct bt_mesh_model *m = root;
-	int depth = 0;
-	/* 'skip' is set to true when we ascend from child to parent node.
-	 * In that case, we want to skip calling the callback on the parent
-	 * node and we don't want to descend onto a child node as those
-	 * nodes have already been visited.
-	 */
-	bool skip = false;
 
-	do {
-		if (!skip &&
-		    cb(m, (uint32_t)depth, user_data) == BT_MESH_WALK_STOP) {
+	for (it = model; it && it->next != model; it = it->next) {
+		if (cb(it, user_data) == BT_MESH_WALK_STOP) {
 			return;
 		}
-#ifdef CONFIG_BT_MESH_MODEL_EXTENSIONS
-		if (!skip && m->extends) {
-			m = m->extends;
-			depth++;
-		} else if (m->flags & BT_MESH_MOD_NEXT_IS_PARENT) {
-			m = m->next;
-			depth--;
-			skip = true;
-		} else {
-			m = m->next;
-			skip = false;
-		}
+	}
 #endif
-	} while (m && depth > 0);
 }
 
 #ifdef CONFIG_BT_MESH_MODEL_EXTENSIONS
-int bt_mesh_model_extend(struct bt_mesh_model *mod,
-			 struct bt_mesh_model *base_mod)
+int bt_mesh_model_extend(struct bt_mesh_model *extending_mod, struct bt_mesh_model *base_mod)
 {
-	/* Form a cyclical LCRS tree:
-	 * The extends-pointer points to the first child, and the next-pointer
-	 * points to the next sibling. The last sibling is marked by the
-	 * BT_MESH_MOD_NEXT_IS_PARENT flag, and its next-pointer points back to
-	 * the parent. This way, the whole tree is accessible from any node.
-	 *
-	 * We add children (extend them) by inserting them as the first child.
-	 */
-	if (base_mod->next) {
-		return -EALREADY;
+	struct bt_mesh_model *a = extending_mod, *b = base_mod, *a_next = a->next,
+			     *b_next = b->next, *it;
```

Comment:
Nitpicking: From the [Linux Kernel codingg style page, part 8](https://www.kernel.org/doc/html/v4.10/process/coding-style.html#commenting):

> use just one data declaration per line (no commas for multiple data declarations). This leaves you room for a small comment on each item, explaining its use.